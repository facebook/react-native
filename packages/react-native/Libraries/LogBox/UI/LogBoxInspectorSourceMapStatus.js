/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import type {PressEvent} from '../../Types/CoreEventTypes';

import Animated from '../../Animated/Animated';
import Easing from '../../Animated/Easing';
import StyleSheet from '../../StyleSheet/StyleSheet';
import Text from '../../Text/Text';
import LogBoxButton from './LogBoxButton';
import * as LogBoxStyle from './LogBoxStyle';
import * as React from 'react';

type Props = $ReadOnly<{
  onPress?: ?(event: PressEvent) => void,
  status: 'COMPLETE' | 'FAILED' | 'NONE' | 'PENDING',
}>;

function LogBoxInspectorSourceMapStatus(props: Props): React.Node {
  const [state, setState] = React.useState({
    animation: null,
    rotate: null,
  });

  React.useEffect(() => {
    if (props.status === 'PENDING') {
      if (state.animation == null) {
        const animated = new Animated.Value(0);
        const animation = Animated.loop(
          Animated.timing(animated, {
            duration: 2000,
            easing: Easing.linear,
            toValue: 1,
            useNativeDriver: true,
          }),
        );
        // $FlowFixMe[incompatible-call]
        setState({
          animation,
          rotate: animated.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '360deg'],
          }),
        });
        animation.start();
      }
    } else {
      if (state.animation != null) {
        state.animation.stop();
        setState({
          animation: null,
          rotate: null,
        });
      }
    }

    return () => {
      if (state.animation != null) {
        state.animation.stop();
      }
    };
  }, [props.status, state.animation]);

  let image;
  let color;
  switch (props.status) {
    case 'FAILED':
      image = require('./LogBoxImages/alert-triangle.png');
      color = LogBoxStyle.getErrorColor(1);
      break;
    case 'PENDING':
      image = require('./LogBoxImages/loader.png');
      color = LogBoxStyle.getWarningColor(1);
      break;
  }

  if (props.status === 'COMPLETE' || image == null) {
    return null;
  }

  return (
    <LogBoxButton
      backgroundColor={{
        default: 'transparent',
        pressed: LogBoxStyle.getBackgroundColor(1),
      }}
      hitSlop={{bottom: 8, left: 8, right: 8, top: 8}}
      onPress={props.onPress}
      style={styles.root}>
      <Animated.Image
        source={image}
        style={[
          styles.image,
          {tintColor: color},
          state.rotate == null || props.status !== 'PENDING'
            ? null
            : {transform: [{rotate: state.rotate}]},
        ]}
      />
      <Text style={[styles.text, {color}]}>Source Map</Text>
    </LogBoxButton>
  );
}

const styles = StyleSheet.create({
  root: {
    alignItems: 'center',
    borderRadius: 12,
    flexDirection: 'row',
    height: 24,
    paddingHorizontal: 8,
  },
  image: {
    height: 14,
    width: 16,
    marginEnd: 4,
    tintColor: LogBoxStyle.getTextColor(0.4),
  },
  text: {
    fontSize: 12,
    includeFontPadding: false,
    lineHeight: 16,
  },
});

export default LogBoxInspectorSourceMapStatus;
