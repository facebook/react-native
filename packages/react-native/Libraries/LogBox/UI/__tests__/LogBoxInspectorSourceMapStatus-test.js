/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall react_native
 */

'use strict';

const render = require('../../../../jest/renderer');
const LogBoxInspectorSourceMapStatus =
  require('../LogBoxInspectorSourceMapStatus').default;
const React = require('react');

// Mock `LogBoxButton` because we are interested in snapshotting the behavior
// of `LogBoxInspectorSourceMapStatus`, not `LogBoxButton`.
jest.mock('../LogBoxButton', () => ({
  __esModule: true,
  default: 'LogBoxButton',
}));

describe('LogBoxInspectorSourceMapStatus', () => {
  it('should render for failed', async () => {
    const output = await render.create(
      <LogBoxInspectorSourceMapStatus onPress={() => {}} status="FAILED" />,
    );

    expect(output).toMatchSnapshot();
  });

  it('should render for pending', async () => {
    const output = await render.create(
      <LogBoxInspectorSourceMapStatus onPress={() => {}} status="PENDING" />,
    );

    expect(output).toMatchSnapshot();
  });

  it('should render null for complete', async () => {
    const output = await render.create(
      <LogBoxInspectorSourceMapStatus onPress={() => {}} status="COMPLETE" />,
    );

    expect(output).toMatchSnapshot();
  });
});
