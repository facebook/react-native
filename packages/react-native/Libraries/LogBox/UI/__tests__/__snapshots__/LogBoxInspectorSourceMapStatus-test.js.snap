// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogBoxInspectorSourceMapStatus should render for failed 1`] = `
<LogBoxButton
  backgroundColor={
    Object {
      "default": "transparent",
      "pressed": "rgba(51, 51, 51, 1)",
    }
  }
  hitSlop={
    Object {
      "bottom": 8,
      "left": 8,
      "right": 8,
      "top": 8,
    }
  }
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 12,
      "flexDirection": "row",
      "height": 24,
      "paddingHorizontal": 8,
    }
  }
>
  <Image
    collapsable={false}
    source={
      Object {
        "testUri": "../Libraries/LogBox/UI/LogBoxImages/alert-triangle.png",
      }
    }
    style={
      Array [
        Object {
          "height": 14,
          "marginEnd": 4,
          "tintColor": "rgba(255, 255, 255, 0.4)",
          "width": 16,
        },
        Object {
          "tintColor": "rgba(243, 83, 105, 1)",
        },
        null,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "fontSize": 12,
          "includeFontPadding": false,
          "lineHeight": 16,
        },
        Object {
          "color": "rgba(243, 83, 105, 1)",
        },
      ]
    }
  >
    Source Map
  </Text>
</LogBoxButton>
`;

exports[`LogBoxInspectorSourceMapStatus should render for pending 1`] = `
<LogBoxButton
  backgroundColor={
    Object {
      "default": "transparent",
      "pressed": "rgba(51, 51, 51, 1)",
    }
  }
  hitSlop={
    Object {
      "bottom": 8,
      "left": 8,
      "right": 8,
      "top": 8,
    }
  }
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 12,
      "flexDirection": "row",
      "height": 24,
      "paddingHorizontal": 8,
    }
  }
>
  <Image
    collapsable={false}
    source={
      Object {
        "testUri": "../Libraries/LogBox/UI/LogBoxImages/loader.png",
      }
    }
    style={
      Object {
        "height": 14,
        "marginEnd": 4,
        "tintColor": "rgba(250, 186, 48, 1)",
        "transform": Array [
          Object {
            "rotate": "0deg",
          },
        ],
        "width": 16,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "fontSize": 12,
          "includeFontPadding": false,
          "lineHeight": 16,
        },
        Object {
          "color": "rgba(250, 186, 48, 1)",
        },
      ]
    }
  >
    Source Map
  </Text>
</LogBoxButton>
`;

exports[`LogBoxInspectorSourceMapStatus should render null for complete 1`] = `null`;
